### PR TITLE
fix Status#as_direct_timeline querying

### PR DESCRIPTION
**Pull request by takayamaki**
related #7614

#7614 is too big.
I feel it is Premature optimization.

This PR is more simply and no adding pertial index and column.

### CODE REVIEW

**Comment by takayamaki** [comment]
explain analyze using my account( fusagiko(at)imastodon.net ).
several execute and shortest time.

## before
```
mastodon_production=> explain analyze SELECT  "statuses"."id", "statuses"."updated_at" FROM "statuses" LEFT OUTER JOIN mentions ON statuses.id = mentions.status_id AND mentions.account_id = 1 LEFT OUTER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE (mentions.account_id = 1 OR statuses.account_id = 1) AND "statuses"."visibility" = 3 AND 1=1 AND "accounts"."silenced" = false ORDER BY "statuses"."id" DESC LIMIT 20;
                                                                                         QUERY PLAN                                                                                         
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=427062.91..427062.96 rows=20 width=16) (actual time=321.116..321.128 rows=20 loops=1)
   ->  Sort  (cost=427062.91..427063.41 rows=202 width=16) (actual time=321.116..321.120 rows=20 loops=1)
         Sort Key: statuses.id DESC
         Sort Method: top-N heapsort  Memory: 25kB
         ->  Nested Loop  (cost=422313.92..427057.53 rows=202 width=16) (actual time=301.851..320.827 rows=844 loops=1)
               ->  Hash Right Join  (cost=422313.63..425922.02 rows=278 width=24) (actual time=301.833..318.037 rows=844 loops=1)
                     Hash Cond: (mentions.status_id = statuses.id)
                     Filter: ((mentions.account_id = 1) OR (statuses.account_id = 1))
                     Rows Removed by Filter: 85239
                     ->  Index Only Scan using index_mentions_on_account_id_and_status_id on mentions  (cost=0.42..2988.87 rows=1770 width=16) (actual time=0.022..1.518 rows=1729 loops=1)
                           Index Cond: (account_id = 1)
                           Heap Fetches: 1404
                     ->  Hash  (cost=420449.52..420449.52 rows=101495 width=24) (actual time=301.641..301.641 rows=86083 loops=1)
                           Buckets: 65536  Batches: 2  Memory Usage: 2862kB
                           ->  Index Only Scan using index_20180106 on statuses  (cost=0.56..420449.52 rows=101495 width=24) (actual time=0.012..267.337 rows=86083 loops=1)
                                 Index Cond: (visibility = 3)
                                 Heap Fetches: 37270
               ->  Index Scan using index_accounts_on_id on accounts  (cost=0.29..4.07 rows=1 width=8) (actual time=0.002..0.002 rows=1 loops=844)
                     Index Cond: (id = statuses.account_id)
                     Filter: (NOT silenced)
 Planning time: 0.547 ms
 Execution time: 321.189 ms
(22 rows)

mastodon_production=>
```
## after
```
mastodon_production=> explain analyze SELECT  "statuses"."id" FROM "statuses" LEFT OUTER JOIN "mentions" ON "mentions"."status_id" = "statuses"."id" WHERE "statuses"."visibility" = 3 AND "mentions"."account_id" = 1 ORDER BY "statuses"."id" DESC LIMIT 20;
                                                                                      QUERY PLAN                                                                                      
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.86..17787.92 rows=20 width=8) (actual time=0.052..0.446 rows=20 loops=1)
   ->  Nested Loop  (cost=0.86..17787.92 rows=20 width=8) (actual time=0.052..0.436 rows=20 loops=1)
         ->  Index Only Scan Backward using index_mentions_on_account_id_and_status_id on mentions  (cost=0.42..2988.87 rows=1770 width=8) (actual time=0.020..0.094 rows=57 loops=1)
               Index Cond: (account_id = 1)
               Heap Fetches: 57
         ->  Index Scan using statuses_pkey on statuses  (cost=0.43..8.35 rows=1 width=8) (actual time=0.005..0.005 rows=0 loops=57)
               Index Cond: (id = mentions.status_id)
               Filter: (visibility = 3)
               Rows Removed by Filter: 1
 Planning time: 0.327 ms
 Execution time: 0.493 ms
(11 rows)

mastodon_production=>


mastodon_production=> explain analyze SELECT  "statuses"."id" FROM "statuses" WHERE "statuses"."visibility" = 3 AND "statuses"."account_id" = 1 ORDER BY "statuses"."id" DESC LIMIT 20;
                                                              QUERY PLAN                                                               
---------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.56..79.95 rows=17 width=8) (actual time=0.019..0.101 rows=20 loops=1)
   ->  Index Only Scan using index_20180106 on statuses  (cost=0.56..79.95 rows=17 width=8) (actual time=0.019..0.092 rows=20 loops=1)
         Index Cond: ((account_id = 1) AND (visibility = 3))
         Heap Fetches: 20
 Planning time: 0.080 ms
 Execution time: 0.118 ms
(6 rows)

mastodon_production=>


mastodon_production=> explain analyze SELECT  "statuses"."id", "statuses"."updated_at" FROM "statuses" LEFT OUTER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE "statuses"."id"IN (xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx) AND 1=1 AND "accounts"."silenced" = false ORDER BY "statuses"."id" DESC LIMIT 20;
                                                                                                                                                                                                                                                                                                                                                                                                            QUERY PLAN                                                                                                                                                                                                                                                                                                                                                                                                             
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.72..454.48 rows=20 width=16) (actual time=0.107..0.107 rows=0 loops=1)
   ->  Nested Loop  (cost=0.72..658.67 rows=29 width=16) (actual time=0.106..0.106 rows=0 loops=1)
         ->  Index Scan Backward using statuses_pkey on statuses  (cost=0.43..337.97 rows=40 width=24) (actual time=0.106..0.106 rows=0 loops=1)
               Index Cond: (id = ANY ('{xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx}'::bigint[]))
         ->  Index Scan using index_accounts_on_id on accounts  (cost=0.29..8.01 rows=1 width=8) (never executed)
               Index Cond: (id = statuses.account_id)
               Filter: (NOT silenced)
 Planning time: 0.332 ms
 Execution time: 0.141 ms
(9 rows)
```

File: spec/models/status_spec.rb [hunk 1]
```diff
@@ -317,38 +317,43 @@
...
 
     it 'does not include direct statuses not mentioning recipient from followed' do
-      expect(@results).to_not include(@followed_direct_status)
+      results = Status.as_direct_timeline(account)
+      expect(results).to_not include(@followed_direct_status)
     end
 
     it 'includes direct statuses mentioning recipient from non-followed' do
       Fabricate(:mention, account: account, status: @not_followed_direct_status)
-      expect(@results).to include(@not_followed_direct_status)
```
Comment (takayamaki) [inline spec/models/status_spec.rb]: `@results` will be created in the before block.
and Mention object is created after the before block is executed.

File: app/models/status.rb [hunk 2]
```diff
@@ -188,10 +191,21 @@ def as_home_timeline(account)
...
+      end
+
+      if since_id.present?
+        dm_to_me = dm_to_me.where(arel_table[:id].gt(since_id))
+        dm_from_me = dm_from_me.where(arel_table[:id].gt(since_id))
+      end
+
+      query = Status.where(id: (dm_to_me + dm_from_me).uniq)
 
       apply_timeline_filters(query, account, false)
```
Comment (tateisu) [inline app/models/status.rb]: This query does  limit → filter → limit.
If filter removes rows, the result row count may be less than expected limit.

example:
_from_me returns 20 rows (clipped by limit)
_to_me returns 4 rows (there is no more row)
merged row is 22 (2 rows removed by uniq)
filter removes 6 rows
result is 16 rows. (there is more rows, but result is less than expected)

more extreme case, you may get 0 rows even if  there is more rows.

File: app/models/status.rb [hunk 3]
```diff
@@ -188,12 +191,24 @@ def as_home_timeline(account)
...
+
+      if since_id.present?
+        dm_to_me = dm_to_me.where(arel_table[:id].gt(since_id))
+        dm_from_me = dm_from_me.where(arel_table[:id].gt(since_id))
+      end
+
+      dm_to_me = apply_timeline_filters(dm_to_me, account, false)
+      dm_from_me = apply_timeline_filters(dm_from_me, account, false)
+
+      Status.where(id: (dm_to_me + dm_from_me).uniq)
```
Comment (takayamaki) [inline app/models/status.rb]: dm_to_me and dm_from_me each gets filterd statuses with limit
-> There are 2 * limit filtered statuses
-> Line 211 gets ORDER BY id DESC LIMIT limit from 2 * limit filtered statuses
-> as_direct_timeline returns filterd statuses with limit correctly

```
mastodon_production=> explain analyze SELECT  "statuses"."id" FROM "statuses" LEFT OUTER JOIN "mentions" ON "mentions"."status_id" = "statuses"."id" LEFT OUTER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE "statuses"."visibility" = 3 AND "mentions"."account_id" = 1 AND 1=1 AND "accounts"."silenced" = false ORDER BY "statuses"."id" DESC LIMIT 20;
                                                                                         QUERY PLAN                                                                                         
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=1.15..17869.62 rows=15 width=8) (actual time=0.080..0.523 rows=20 loops=1)
   ->  Nested Loop  (cost=1.15..17869.62 rows=15 width=8) (actual time=0.078..0.509 rows=20 loops=1)
         ->  Nested Loop  (cost=0.86..17787.92 rows=20 width=16) (actual time=0.071..0.427 rows=20 loops=1)
               ->  Index Only Scan Backward using index_mentions_on_account_id_and_status_id on mentions  (cost=0.42..2988.87 rows=1770 width=8) (actual time=0.021..0.105 rows=60 loops=1)
                     Index Cond: (account_id = 1)
                     Heap Fetches: 60
               ->  Index Scan using statuses_pkey on statuses  (cost=0.43..8.35 rows=1 width=16) (actual time=0.005..0.005 rows=0 loops=60)
                     Index Cond: (id = mentions.status_id)
                     Filter: (visibility = 3)
                     Rows Removed by Filter: 1
         ->  Index Scan using index_accounts_on_id on accounts  (cost=0.29..4.07 rows=1 width=8) (actual time=0.003..0.003 rows=1 loops=20)
               Index Cond: (id = statuses.account_id)
               Filter: (NOT silenced)
 Planning time: 0.508 ms
 Execution time: 0.559 ms
(15 rows)

mastodon_production=> explain analyze SELECT  "statuses"."id" FROM "statuses" LEFT OUTER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE "statuses"."visibility" = 3 AND "statuses"."account_id" = 1 AND 1=1 AND "accounts"."silenced" = false ORDER BY "statuses"."id" DESC LIMIT 20;
                                                                   QUERY PLAN                                                                    
-------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.85..88.47 rows=17 width=8) (actual time=0.043..1.021 rows=20 loops=1)
   ->  Nested Loop  (cost=0.85..88.47 rows=17 width=8) (actual time=0.042..1.011 rows=20 loops=1)
         ->  Index Only Scan using index_20180106 on statuses  (cost=0.56..79.95 rows=17 width=16) (actual time=0.031..0.976 rows=20 loops=1)
               Index Cond: ((account_id = 1) AND (visibility = 3))
               Heap Fetches: 20
         ->  Materialize  (cost=0.29..8.31 rows=1 width=8) (actual time=0.001..0.001 rows=1 loops=20)
               ->  Index Scan using index_accounts_on_id on accounts  (cost=0.29..8.31 rows=1 width=8) (actual time=0.008..0.009 rows=1 loops=1)
                     Index Cond: (id = 1)
                     Filter: (NOT silenced)
 Planning time: 0.170 ms
 Execution time: 1.054 ms
(11 rows)

imastodon=> explain analyze SELECT  "statuses"."id", "statuses"."updated_at" FROM "statuses" WHERE "statuses"."id"IN (xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx) ORDER BY "statuses"."id" DESC LIMIT 20;
                                                                                                                                                                                                                                                                                                                                                                                                         QUERY PLAN                                                                                                                                                                                                                                                                                                                                                                                                          
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.43..169.20 rows=20 width=16) (actual time=0.105..0.105 rows=0 loops=1)
   ->  Index Scan Backward using statuses_pkey on statuses  (cost=0.43..337.97 rows=40 width=16) (actual time=0.104..0.104 rows=0 loops=1)
         Index Cond: (id = ANY ('{xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx}'::bigint[]))
 Planning time: 0.122 ms
 Execution time: 0.128 ms
(5 rows)
```

**Comment by tateisu** [comment]
I thought about it now.

This TL is close to home + notification rather  than public.
There is no need to filter silence in this TL?
Regardless of the presence or absence of silence, I should be able to read my remarks and mentions for myself.

    account_silencing_filter: certainly unnecessary
    not_in_filtered_languages: unnecessary
    not_domain_blocked_by_account: unnecessary
    not_excluded_by_account: REQUIRED?

**Comment by takayamaki** [comment]
#7614 merged without adding pertial index and column.
Then, I close this PR.

**Pull request closed without merging**